include/observable.h

```diff
@@ -0,0 +1,271 @@
+
+#ifndef OBSERVABLE_H
+#define OBSERVABLE_H
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <wx/config.h>
+#include <wx/event.h>
+
+/** Return address as printable string. */
+std::string ptr_key(const void* ptr);
+
+
+class ObservedVar;
+class ObservedVarListener;
+
+
+/**
+ *  Private helper class. Basically a singleton map of listeners where
+ *  singletons are managed by key, one for each key value.
+ */
+class SingletonVar {
+
+private:
+  static SingletonVar* getInstance(const std::string& key);
+  std::map<wxEvtHandler*, wxEventType> listeners;
+
+  SingletonVar() {}
+  SingletonVar(const SingletonVar&);    // not implemented
+  void operator=(const SingletonVar&);  // not implemented
+
+  friend class ObservedVar;
+};
+
+
+/**  The observable notify/listen basic nuts and bolts.  */
+class ObservedVar {
+public:
+  ObservedVar(const std::string& key)
+      : singleton(SingletonVar::getInstance(key)) {}
+
+  /** Notify all listeners about variable change. */
+  virtual const void notify();
+
+  /** Remove window from list of listeners, return true if listener exists */
+  bool unlisten(wxEvtHandler* listener);
+
+
+  /** Get listening object which unlisten() on destruction. */
+  typedef std::unique_ptr<ObservedVarListener> Listener;
+  Listener get_listener(wxEvtHandler* handler, wxEventType ev);
+
+
+protected:
+  /**
+   * Notify all listeners: send them a 'type' wxCommandEvent message
+   * as defined by listen() with optional data available using GetString()
+   * and/or GetClientData().
+   */
+  const void notify(const std::string& s, void* client_data);
+
+private:
+  /** Set object to send ev_type to listener on variable changes. */
+  void listen(wxEvtHandler* listener, wxEventType ev_type);
+
+  SingletonVar* const singleton;
+
+  friend class ObservedVarListener;
+};
+
+
+/**
+ *  Keeps listening over it's lifespan, removes itself on destruction.
+ *  Basically private, only available through ObservedVar::get_listener().
+ */
+class ObservedVarListener {
+public:
+  virtual ~ObservedVarListener() { var.unlisten(listener); };
+
+private:
+  /** Set object to send ev_type to window on variable changes. */
+  ObservedVarListener(ObservedVar v, wxEvtHandler* w, wxEventType ev)
+    : var(v), listener(w)  { var.listen(listener, ev); }
+
+  ObservedVar var;
+  wxEvtHandler* listener;
+
+  friend class ObservedVar;
+};
+
+
+/**
+ * Generic event handling between MVC Model and Controller based on a
+ * shared EventVar variable
+ *
+ *  Model usage:
+ *
+ *   class Model: ...
+ *   public:
+ *     EventVar change;
+ *
+ *     void some_method() {
+ *       ...
+ *       change.notify("new value")
+ *     }
+ *
+ *  Controller/GUI usage:
+ *
+ *    class Gui: public wxEvtHandler {
+ *    public:
+ *      Gui:Gui(Model& model) {
+ *        wxDEFINE_EVENT(EVT_FOO, wxCommandEvent);
+ *        change_listener = model.change.get_listener(this, EVT_FOO);
```

Comment:
A more typical idiom would be to have a stack-allocated RAII helper that does all the work in its constructor. This line would become something like:

```
ObservedVar::Listener change_listener(model.change);
```

Reply:
This is the core thing here. You are of course right, also I have adapted to the overall pointers-everywhere idiom this code lives with.  Will fix.

Reply:
hm... looking more into this it is perhaps not the best option. The problem with a stack allocated variable is that has to be constructed like

    wxDEFINE_EVENT(EVT_FOO, wxCOMMAND_EVENT)
    ObservableListener(..., EVT_FOO);

This is not easily done in the initial declaration. This then requires a default CTOR which is not well defined.  It also pushes a need for a default CTOR to ObservedVar, also not really defined. 

Perhaps it's better to leave as-is which guarantees that all Listeners have a sane state, possibly 0. If this indeed is the case it's basically just luck, I have not considered the alternative approach as I should have.

If you have time: thoughts?

Reply:
Here's a prototype of how to do it: https://github.com/resistor/OpenCPN/commit/a75366b4582450e3e386f5f216990e2c3aa6da15

The trick is that you need to initialize the ObservedVariableListener in the class declaration, or in the constructor's initializer list. As long as it's initialized in one of those two places, a default constructor will not be required. Also, you can use deleted-declarations to get better error messages about this.

In my prototype above I used in-class-declaration because it's my typical preference, but I did have to pull some declarations into the header to make it work. Doing this in the constructor initializer list would avoid that, but is a bit less ergonomic overall.

Reply:
Indeed. However, it's not clear-cut: there is a prize to pay in the form of yet another module variable. That is, to initiate the listener on module level the variable to listen to must also exist on module level. In the case of events, this is more or less how things should be used. But in other cases, the actual variables (config_var, global_var, ...) are today typically declared in a function where they are used. 

It also means that the code for listening to an event could be kept together in one place, rather than spread out in several places.

In short, after digging into this I'm not really convinced  that using stack-allocated variables is a better approach in this specific case That is certainly not to say you are wrong in general.

Reply:
I can almost certainly eliminate the need for a second variable (and it's member-scoped, not module-scoped). I was just trying to be minimally invasive to your existing change.

Reply:
Perhaps. But the listener CTOR needs a variable (really, the key) so it will, like you say, be pretty invasive. And there will still be the effect that the code for listening to a change will be spread out all over the class. This is not ideal, there is plenty of far too big classes in use -- spreading things in those is a clear drawback.

So, while I initially embraced the use of variables rather than pointers I now after some digging is leaning to that current solution actually is good.  Standard patterns are not always the best of all worlds when facing pesky macros and classes thousands of lines large. 

perhaps we should close this thread...

Reply:
I have pushed a new commit using stack-based listeners. The basic trick is to actually provide a default CTOR. With this, the module listener instances can be given a new value when actually starting to listen. This can be done in the same place as today.

The get_listener() method is not necessary any more, it's just a shorthand for the somewhat clumsy CTOR.  Left as is, seems convenient.

There is no free lunch.  The prize paid here is a new raw pointer used internally. This points to the variable listening, and might very well become stale. More work under way...

Reply:
OK, final version for now.  This basically adds some code to the listener so it can cope with the empty, default-constructed  items. The raw pointer is retired, using the key for the ObservedVar instead, re-creating the variable form key when needed.

I actually think this should be more robust the the previous implementation.  As of previous version, it allows handling all listening stuff in one place besides the listener declaration.

Reply:
Closing this thread. It seems like you are heading for another kind functionality than this PR, see [below](https://github.com/OpenCPN/OpenCPN/pull/2607#issuecomment-1123377213)